src/main/java/com/monumental/services/MonumentService.java

```diff
@@ -19,34 +17,59 @@ public MonumentService(SessionFactoryService sessionFactoryService) {
-     * TODO: Search other fields besides title
-     * @param query The search string
-     * @return      Matching Monuments by title
+     * Creates a Full Text Search on the Monument's title and artist fields, and adds them to your CriteriaQuery
+     * @param builder       Your CriteriaBuilder
+     * @param query         Your CriteriaQuery
+     * @param root          The root associated with your CriteriaQuery
+     * @param searchQuery   The string to search both fields for
+     */
+    private void buildFTSQuery(CriteriaBuilder builder, CriteriaQuery query, Root root, String searchQuery) {
+        query.where(
+            builder.or(
+                builder.equal(builder.function("fts", Boolean.class, root.get("title"), builder.literal(searchQuery)), true),
+                builder.equal(builder.function("fts", Boolean.class, root.get("artist"), builder.literal(searchQuery)), true)
+            )
+        );
+    }
+
+    /**
+     * Uses the FTS function as well as any filtering or pagination provided to search for matching Monuments
      */
     @SuppressWarnings("unchecked")
-    public List<Monument> search(String query) {
-        Session session = this.sessionFactoryService.getFactory().openSession();
-        Transaction transaction = null;
-        List<Monument> records;
-
-        try {
-            transaction = session.beginTransaction();
-            // The "fts" function is defined by the FTSFunction and CustomPostgreSQL9Dialect classes
-            Query q = session.createQuery(
-                    "select m from Monument m where fts(m.title, :query) = true"
-            ).setParameter("query", query);
-            records = q.list();
-            transaction.commit();
-            session.close();
-        } catch (HibernateException e) {
-            if (transaction != null) {
-                transaction.rollback();
-            }
-            session.close();
-            throw e;
+    public List<Monument> search(String searchQuery, String page, String limit) {
+
+        CriteriaBuilder builder = this.getCriteriaBuilder();
+        CriteriaQuery<Monument> query = this.createCriteriaQuery(builder, false);
+        Root<Monument> root = this.createRoot(query);
+        root.fetch("tags", JoinType.LEFT);
+        query.select(root).distinct(true);
+
+        // TODO: Query for tags
+        // TODO: Query for location
+        // TODO: Filters
+
+        if (searchQuery != null) {
+            this.buildFTSQuery(builder, query, root, searchQuery);
         }
 
-        return new ArrayList<>(new HashSet<>(records));
+        List<Monument> monuments = this.getWithCriteriaQuery(query, Integer.parseInt(limit), (Integer.parseInt(page)) - 1);
+
+        return monuments;
+    }
+
+    /**
+     * Count the total number of results for a FTS search
+     * TODO: Be sure to update this whenever the search function changes so that they stay in sync
+     */
+    public Integer countSearchResults(String searchQuery) {
```

Comment:
Where is this used? Is there missing front-end code that uses this? (I'm assuming so).

Reply:
Sorry, I meant the Controller method that calls this, but same thing.

Reply:
Yes it's used by the search page. I wanted to try to split back & front end pieces so that the PRs are a little more manageable